[quant][fx] Change prepare_fx and convert_fx to use GraphModule

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92842

Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can use GraphModule directly, this
is useful for quantization in pytorch 2.0 flow, if other transformations are using GraphModule as well, the quantization passes will be composable with them

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42979722](https://our.internmc.facebook.com/intern/diff/D42979722)

jerryzh168 committed:
[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input

Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

qihqi commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -450,11 +450,12 @@ def _restore_state(
-    prepare_custom_config: PrepareCustomConfig = observed._prepare_custom_config  # type: ignore[assignment]
-    node_name_to_scope: Dict[str, Tuple[str, type]] = observed._node_name_to_scope  # type: ignore[assignment]
-    observed_node_names: Set[str] = observed._observed_node_names  # type: ignore[assignment]
+    # assert _is_observed_module(observed), \
```
Can we have a dataclass
```python
@dataclass
class QuantizationConfig: # or other name
  prepare_custom_config: ...
  node_to_scope: ...
  observed_node_names:...
  ...
```
basically all those individual fields that need to be present for an Observed/Fused/Quantized module.

Then have one instance set into graph_module.meta['_quantized_config']

The pros is that 1. it's all in the same place and we know presence of it by checking one key in the meta dict.
and 2. code change is easier: for example: add observed = observed.meta before L454 then dont need to change anything after it as we kept the dot syntax with the dataclass.

jerryzh168 replied:
sure, sounds good

jerryzh168 replied:
actually, we will also have some preserved attributes as well requested by users, we still need to "pollute" the keys of model.meta..

qihqi replied:
well, you can have a user_preserved_attr dict in the QuantizationConfig dataclass and use that instead?

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -336,10 +336,29 @@ def should_skip_lowering(op: torch.fx.node.Node, qconfig_map: Dict[str, QConfigA
     torch.mul: torch.ops.quantized.mul_relu,
 }
 
+def _save_packed_weight(self, destination, prefix, keep_vars):
```
just curious, what's the motivation for this?  I would have expected the packed weight to exist in memory, but the unpacked version to be stored on disk.  Is that not how it works anymore?

jerryzh168 replied:
the motivation for this is when saving things to state_dict, by default only parameters and buffers are saved: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1711-L1716, but packed_weight is not a parameter or buffer

jerryzh168 replied:
looks like we added a `get_extra_state` and `set_extra_state` to `nn.Module`, maybe we could use that as well..

vkuzo replied:
oh, my question is why this is saving the packed weight instead of saving the unpacked version?  I thought in all of our workflows the unpacked version is for saving to disk (for historical reasons, for fbgemm vs qnnpack compatibility), and the in-memory version is the only thing that is packed.

or is this code doing that and I'm just not reading it correctly?

jerryzh168 replied:
yeah, what you described is happening as part of the serialization for the packed_weight object itself `torch._C.ScriptObject`, this object is put into state_dict, but the way we serialize this object is to unpack it and save the Tensors

vkuzo replied:
got it, thanks for explaining, makes sense!

vkuzo commented on torch/ao/quantization/fx/graph_module.py:
```diff
@@ -55,7 +55,7 @@ def __deepcopy__(self, memo):
 
 def _is_observed_module(module: Any) -> bool:
-    return isinstance(module, ObservedGraphModule)
+    return hasattr(module, "meta") and "_observed_graph_module_attrs" in module.meta
```
just curious, can we delete FusedGraphModule too in the future?

jerryzh168 replied:
yes, I'm planning to delete all GraphModules in this file

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42979722).

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42979722](https://our.internmc.facebook.com/intern/diff/D42979722)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42979722](https://our.internmc.facebook.com/intern/diff/D42979722)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][bc-breaking] Change prepare_fx and convert_fx to preserve the GraphModule type of input"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can preserve the original GraphModule type now, this
is useful for quantization in pytorch 2.0 flow when we get a different type of GraphModule from some other program capture mechanisms

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42979722](https://our.internmc.facebook.com/intern/diff/D42979722)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Change prepare_fx and convert_fx to use GraphModule"


Summary:
Previously prepare_fx returns an ObservedGraphModule and convert_fx returns a QuantizedGraphModule,
this is to preserve the attributes since torch.fx.GraphModule did not preserve them, after https://github.com/pytorch/pytorch/pull/92062
we are preserving `model.meta`, so we can store the attributes in model.meta now to preserve them.

With this, we don't need to create a new type of GraphModule in these functions and can use GraphModule directly, this
is useful for quantization in pytorch 2.0 flow, if other transformations are using GraphModule as well, the quantization passes will be composable with them

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D42979722](https://our.internmc.facebook.com/intern/diff/D42979722)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42979722).

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorch merge -f "landing in oss first"

jerryzh168 commented:
@pytorchbot merge -f "landing in oss first"

jerryzh168 commented:
closing this PR since we have internal changes and need to land from phabricator
